File: go/ledger/token.go
package ledger

import (
	"github.com/meshtrade/api/go/num"
	"github.com/meshtrade/api/go/search"
	"github.com/shopspring/decimal"
)

func NewUndefinedToken() *Token {
	return &Token{
		Code:    "-",
		Issuer:  "-",
		Network: Network_UNDEFINED_NETWORK,
	}
}

func (t *Token) NewAmountOf(value decimal.Decimal) *Amount {
	switch t.Network {
	case Network_STELLAR_NETWORK:
		return &Amount{
			Value: num.FromShopspringDecimal(value.Truncate(7)),
			Token: t,
		}

	default:
		return &Amount{
			Value: num.FromShopspringDecimal(value),
			Token: t,
		}
	}
}

func (t *Token) IsEqualTo(t2 *Token) bool {
	return t.Code == t2.Code &&
		t.Issuer == t2.Issuer &&
		t.Network == t2.Network
}

func (t *Token) IsUndefined() bool {
	return t == nil || NewUndefinedToken().IsEqualTo(t)
}

func (t *Token) ToCriterion() []*search.Criterion {
	return []*search.Criterion{
		search.NewTextExactCriterion("token.code", t.Code),
		search.NewTextExactCriterion("token.issuer", t.Issuer),
		search.NewUint32ExactCriterion("token.network", uint32(t.Network)),
	}
}

File: ts/src/ledger/token.ts
import { newTextExactCriterion, newUint32ExactCriterion } from "../search";
import { Criterion } from "../search/criterion_pb";
import { Network } from "./network_pb";
import { Token } from "./token_pb";

export function criteriaFromToken(token?: Token): Criterion[] {
  return [
    newTextExactCriterion("token.code", token?.code ?? ""),
    newTextExactCriterion("token.issuer", token?.issuer ?? ""),
    newUint32ExactCriterion(
      "token.network",
      token?.network ?? Network.UNDEFINED_NETWORK,
    ),
  ];
}

export function tokensAreEqual(t?: Token, t2?: Token): boolean {
  return (
    (t?.code ?? "") === (t2?.code ?? "") &&
    (t?.issuer ?? "") === (t2?.issuer ?? "") &&
    (t?.network ?? Network.UNDEFINED_NETWORK) ===
    (t2?.network ?? Network.UNDEFINED_NETWORK)
  );
}

export function tokenIsUndefined(t?: Token): boolean {
  if (!t) {
    return true;
  }
  return (
    (t.code === "-" || t.code === "") &&
    (t.issuer === "-" || t.issuer === "") &&
    (t.network === Network.NULL_NETWORK ||
      t.network === Network.UNDEFINED_NETWORK)
  );
}

export function tokenToFilter(t?: Token): Criterion[] {
  return [
    newTextExactCriterion("token.code", t?.code ?? ""),
    newTextExactCriterion("token.issuer", t?.issuer ?? ""),
    newUint32ExactCriterion(
      "token.network",
      t?.network ?? Network.UNDEFINED_NETWORK,
    ),
  ];
}

export function tokenToString(t?: Token): string {
  return `${t?.code ?? "-"} by ${t?.issuer ?? "-"} on ${t?.network ?? Network.UNDEFINED_NETWORK}`;
}
